Public/Src/Cache/ContentStore/Distributed/NuCache/EventStreaming/ContentLocationEventData.cs

```diff
@@ -310,6 +318,55 @@ public override int GetHashCode()
         {
             return (base.GetHashCode(), AccessTime.GetHashCode()).GetHashCode();
         }
+    }
+
+    /// <nodoc />
+    public sealed class AddContentLocationWithoutTouchingEventData : ContentLocationEventData
+    {
+        /// <summary>
+        /// A list of content sizes associated with <see cref="ContentLocationEventData.ContentHashes"/>.
+        /// </summary>
+        public IReadOnlyList<long> ContentSizes { get; }
+
+        /// <nodoc />
+        public AddContentLocationWithoutTouchingEventData(MachineId sender, IReadOnlyList<ShortHashWithSize> addedContent)
+            : this(sender, addedContent.SelectList(c => c.Hash), addedContent.SelectList(c => c.Size))
+        {
+        }
+
+        /// <nodoc />
+        public AddContentLocationWithoutTouchingEventData(MachineId sender, IReadOnlyList<ShortHash> contentHashes, IReadOnlyList<long> contentSizes)
+            : base(EventKind.AddLocationWithoutTouching, sender, contentHashes)
+        {
+            Contract.Requires(contentSizes != null);
+            Contract.Requires(contentSizes.Count == contentHashes.Count);
+
+            ContentSizes = contentSizes;
+        }
 
+        /// <inheritdoc />
+        public override string ToString()
+        {
+            var sb = new StringBuilder();
+            for (int i = 0; i < ContentHashes.Count; i++)
+            {
+                sb.Append($"Hash={ContentHashes[i]}, Size={ContentSizes[i]}");

```

Comment:
As written, these are going to run together, e.g. "Hash=ABC, Size=1Hash=DEF, Size=2Hash=123, Size=3"